README.rst
apache-manager: Monitor and control Apache web server workers from Python
=========================================================================

.. image:: https://travis-ci.org/xolox/python-apache-manager.svg?branch=master
   :target: https://travis-ci.org/xolox/python-apache-manager

.. image:: https://coveralls.io/repos/xolox/python-apache-manager/badge.svg?branch=master
   :target: https://coveralls.io/r/xolox/python-apache-manager?branch=master

The `apache-manager` package helps to collect monitoring metrics_ about the
`Apache web server`_ and kill worker processes that exceed resource thresholds.
It works by parsing the plain text and HTML status pages generated by Apache's
mod_status_ module. It's currently tested on cPython 2.7, 3.5, 3.6, 3.7, 3.8
and PyPy. It was developed for and expects to be run on Linux systems.

.. contents::
   :local:

Status
------

On the one hand the `apache-manager` package was developed based on quite a few
years of experience interfacing with (and monitoring) the `Apache web server`_
using the interfaces provided by Apache itself (mod_status_) as well as the
Linux operating system interfaces (``/proc``).

On the other hand the Python package itself is quite new: it was developed
between June and September of 2015.

Then again, I'm religious about test coverage, so there's that :-).

Despite the complete package being covered by automated tests I'm not (yet)
committing to backwards compatibility because I first want to run this package
in a production (like) environment for some time, to get a feeling for how
robust it is. I also need to evaluate how easy it is to integrate the current
solution into monitoring systems (that was a goal, but I don't always achieve
my goals on the first attempt :-).

Installation
------------

The `apache-manager` package is available on PyPI_ which means installation
should be as simple as:

.. code-block:: console

   $ pip install apache-manager

There's actually a multitude of ways to install Python packages (e.g. the `per
user site-packages directory`_, `virtual environments`_ or just installing
system wide) and I have no intention of getting into that discussion here, so
if this intimidates you then read up on your options before returning to these
instructions ;-).

Usage
-----

There are two ways to use the `apache-manager` package: As the command line
program ``apache-manager`` and as a Python API. For details about the Python
API please refer to the API documentation available on `Read the Docs`_. The
command line interface is described below.

Command line
~~~~~~~~~~~~

.. A DRY solution to avoid duplication of the `apache-manager --help' text:
..
.. [[[cog
.. from humanfriendly.usage import inject_usage
.. inject_usage('apache_manager.cli')
.. ]]]

**Usage:** `apache-manager [OPTIONS]`

Command line interface to monitor the Apache web server and kill worker
processes that exceed resource thresholds. When no options are given the
server metrics and memory usage of workers are printed to the terminal.

**Supported options:**

.. csv-table::
   :header: Option, Description
   :widths: 30, 70


   "``-w``, ``--watch``","This option causes the Apache manager to redraw the collected metrics once
   every 10 seconds in a ""top"" like interface until interrupted using ""q"" (for
   quite) or Control-C."
   "``-a``, ``--max-memory-active=SIZE``","Kill active Apache workers that are using more memory than specified by the
   ``SIZE`` argument. ``SIZE`` is expected to be a human readable memory size like 50K
   (50 kilobytes), 42M (42 megabytes), 2G (2 gigabytes), etc."
   "``-i``, ``--max-memory-idle=SIZE``","Kill Apache workers that are using more memory than specified by the ``SIZE``
   argument (see ``--max-memory-active`` for acceptable values of ``SIZE``)."
   "``-t``, ``--max-ss``, ``--max-time=TIMESPAN``","Kill Apache workers whose ""time since the beginning of the most recent
   request"" is greater than specified by the ``TIMESPAN`` argument. ``TIMESPAN`` is
   expected to be a human readable timespan like 2s (2 seconds), 3m (3
   minutes), 5h (5 hours), 2d (2 days), etc."
   "``-f``, ``--data-file=PATH``","Change the pathname of the file where the Apache manager stores monitoring
   metrics after every run. Defaults to ""/tmp/apache-manager.txt""."
   "``-z``, ``--zabbix-discovery``","Generate a JSON fragment that's compatible with the low-level discovery
   support in the Zabbix monitoring system. With the right template in place
   this enables the Zabbix server to discover the names of the WSGI process
   groups that are active on any given server. This makes it possible to
   collect and analyze the memory usage of specific WSGI process groups."
   "``-n``, ``--dry-run``, ``--simulate``",Don't actually kill any Apache workers.
   "``-v``, ``--verbose``",Increase verbosity (can be repeated).
   "``-q``, ``--quiet``",Decrease verbosity (can be repeated).
   "``-h``, ``--help``",Show this message and exit.

.. [[[end]]]

Contact
-------

The latest version of `apache-manager` is available on PyPI_ and GitHub_. The
documentation is hosted on `Read the Docs`_ and includes a changelog_. For bug
reports please create an issue on GitHub_. If you have questions, suggestions,
etc. feel free to send me an e-mail at `peter@peterodding.com`_.

License
-------

This software is licensed under the `MIT license`_.

© 2020 Peter Odding.


.. External references:
.. _Apache web server: https://en.wikipedia.org/wiki/Apache_HTTP_Server
.. _changelog: https://apache-manager.readthedocs.io/en/latest/changelog.html
.. _documentation: https://apache-manager.readthedocs.io
.. _GitHub: https://github.com/xolox/python-apache-manager
.. _metrics: https://en.wikipedia.org/wiki/System_monitoring
.. _MIT license: http://en.wikipedia.org/wiki/MIT_License
.. _mod_status: http://httpd.apache.org/docs/current/mod/mod_status.html
.. _per user site-packages directory: https://www.python.org/dev/peps/pep-0370/
.. _peter@peterodding.com: peter@peterodding.com
.. _PyPI: https://pypi.python.org/pypi/apache-manager
.. _Read the Docs: https://apache-manager.readthedocs.io
.. _virtual environments: http://docs.python-guide.org/en/latest/dev/virtualenvs/


CHANGELOG.rst
Changelog
=========

The purpose of this document is to list all of the notable changes to this
project. The format was inspired by `Keep a Changelog`_. This project adheres
to `semantic versioning`_.

.. contents::
   :local:

.. _Keep a Changelog: http://keepachangelog.com/
.. _semantic versioning: http://semver.org/

`Release 2.2`_ (2020-02-29)
---------------------------

- Make the command line interface compatible with the ``python -m
  apache_manager …`` command.

- Added :ref:`Zabbix integration for apache-manager` to the online
  documentation.

.. _Release 2.2: https://github.com/xolox/python-apache-manager/compare/2.1...2.2

`Release 2.1`_ (2020-02-27)
---------------------------

This release changes how logging is done:

- When workers are killed this is now logged at the custom level ``NOTICE``
  where previously the level ``INFO`` was used (``NOTICE`` sits between
  ``INFO`` and ``WARNING``). The custom log level is implemented by
  :pypi:`verboselogs`.

- System logging has been reduced so that only killed workers, warnings and
  errors are logged. This is because when ``apache-manager`` is being run from
  a high frequency cron job it shouldn't spam the system logs.

Additionally some changes were made to the test suite:

- Use proper skipping so that :pypi:`pytest` is aware of skipped tests.

- Use :pypi:`pytest-rerunfailures` to automate high level retrying of failed
  tests (duct taping away flaky tests). I'd love to make the test suite more
  robust in the near future but lack the time to do so now.

.. _Release 2.1: https://github.com/xolox/python-apache-manager/compare/2.0...2.1

`Release 2.0`_ (2020-02-26)
---------------------------

**Backwards incompatible changes:**

- Drop support for Python 2.6 and 3.4, start testing on Python 3.8.

- Explicit command line options are now required to enable metrics collection
  and killing of workers. This was prompted by the following awkward
  interaction:

  - The command line interface was initially designed such that killing of
    workers was enabled when thresholds were given as command line options.

  - Since then support for configuration files was added, and given the
    presence of a configuration file thresholds would always be set so
    killing would happen implicitly and unconditionally.

  To solve these `explicit is better than implicit`_ contradictions all in one
  go I decided to make a backwards incompatible change to the command line
  interface, where both of the actions described above now need to be requested
  using command line options.

- Parsing of sizes now uses binary multiples of bytes (base-2) for ambiguous
  unit symbols and names whereas previously decimal multiples of bytes
  (base-10) were used.

- The custom initializer on the main :class:`~apache_manager.ApacheManager`
  class was removed because it was a historical artefact whose significance was
  lost in time.

**Other significant changes:**

- Make the :attr:`~apache_manager.ApacheManager.hanging_worker_threshold`
  option configurable.

- Add support for configuration files to configure killing of workers (for
  details see the :attr:`~apache_manager.ApacheManager.config_loader`
  property).

- Expose native and foreign worker count in data file to enable monitoring that
  detects configuration issues (like the native worker count being lower than
  the foreign worker count, causing the native workers to become saturated).

**Miscellaneous changes:**

- Document that Linux is required (`#2`_).
- Improve string representation of WSGI workers.
- Use Python 3 for local development in ``Makefile``.

.. _Release 2.0: https://github.com/xolox/python-apache-manager/compare/1.2...2.0
.. _explicit is better than implicit: https://www.python.org/dev/peps/pep-0020/#the-zen-of-python
.. _#2: https://github.com/xolox/python-apache-manager/issues/2

`Release 1.2`_ (2019-03-28)
---------------------------

- Added Python 3.6 and 3.7 to test suite and documented support for them (based
  on the fact that the test suite passes).

- Bug fix to improve compatibility with newer Apache versions:

  In Ubuntu 18.04 the plain text server status page response contains multiple
  uptime entries and because the regular expressions used by apache-manager
  weren't anchored to the start of the line, this new status page contents
  confused apache-manager.

  On Ubuntu 14.04:

  .. code-block:: console

     $ curl -s http://localhost/server-status?auto | grep -i uptime
     Uptime: 96606

  On Ubuntu 18.04:

  .. code-block:: console

     $ curl -s http://localhost/server-status?auto | grep -i uptime
     ServerUptimeSeconds: 5163
     ServerUptime: 1 hour 26 minutes 3 seconds
     Uptime: 5163

- Include documentation in source distributions (``MANIFEST.in``).

- Changed Sphinx documentation theme (to the 'nature' theme).

- Added license=MIT to ``setup.py`` script.

.. _Release 1.2: https://github.com/xolox/python-apache-manager/compare/1.1...1.2

`Release 1.1`_ (2017-03-06)
---------------------------

- Added hanging worker detection based on hard coded five minute threshold (to
  be made configurable in a future release).

- Made ``test_refresh()`` compatible with Ubuntu 16.04:

  I've just upgraded my personal and work laptops to Ubuntu 16.04 and noticed
  that several tests have started failing. Most noticeably the server uptime
  reported on the status page is no longer updated consistently. I'm changing
  this test to check a different status page item which should be more
  reliable.

- Made worker kill tests compatible with Ubuntu 16.04 (Apache 2.4).

.. _Release 1.1: https://github.com/xolox/python-apache-manager/compare/1.0...1.1

`Release 1.0`_ (2017-02-15)
---------------------------

- Refactor ``WorkerStatus`` class to properly use :pypi:`property-manager`.

  Strictly speaking this change set breaks backwards compatibility, however 99%
  percent of the functionality is the same, I've mostly just changed a whole

apache_manager/cli.py
# Monitor and control Apache web server workers from Python.
#
# Author: Peter Odding <peter@peterodding.com>
# Last Change: February 27, 2020
# URL: https://apache-manager.readthedocs.io

"""
Usage: apache-manager [OPTIONS]

Command line interface to monitor the Apache web server and kill worker
processes that exceed resource thresholds. When no options are given the
server metrics and memory usage of workers are printed to the terminal.

Supported options:

  -c, --collect-metrics

    Collect monitoring metrics and store them in a text file to be read
    by a monitoring system like Zabbix. See also the --data-file option.

  -k, --kill-workers

    Kill Apache workers exceeding the thresholds given by --max-memory-active,
    --max-memory-idle and --max-time. These thresholds can also be defined in
    configuration files, please refer to the online documentation for details.
    See also the --dry-run option.

  -w, --watch

    This option causes the Apache manager to redraw the collected metrics once
    every 10 seconds in a `top' like interface until interrupted using `q' (for
    quite) or Control-C.

  -a, --max-memory-active=SIZE

    Kill active Apache workers that are using more memory than specified by the
    SIZE argument. SIZE is expected to be a human readable memory size like 50K
    (50 kilobytes), 42M (42 megabytes), 2G (2 gigabytes), etc.

  -i, --max-memory-idle=SIZE

    Kill Apache workers that are using more memory than specified by the SIZE
    argument (see --max-memory-active for acceptable values of SIZE).

  -t, --max-ss, --max-time=TIMESPAN

    Kill Apache workers whose "time since the beginning of the most recent
    request" is greater than specified by the TIMESPAN argument. TIMESPAN is
    expected to be a human readable timespan like 2s (2 seconds), 3m (3
    minutes), 5h (5 hours), 2d (2 days), etc.

  -T, --hanging-worker-threshold=TIMESPAN

    Change the number of seconds before an active worker is considered hanging
    to TIMESPAN (see --max-time for acceptable values of TIMESPAN).

  -f, --data-file=PATH

    Change the pathname of the file where the Apache manager stores monitoring
    metrics after every run. Defaults to `/tmp/apache-manager.txt'.

  -z, --zabbix-discovery

    Generate a JSON fragment that's compatible with the low-level discovery
    support in the Zabbix monitoring system. With the right template in place
    this enables the Zabbix server to discover the names of the WSGI process
    groups that are active on any given server. This makes it possible to
    collect and analyze the memory usage of specific WSGI process groups.

  -n, --dry-run, --simulate

    Don't actually kill any Apache workers.

  -v, --verbose

    Increase verbosity (can be repeated).

  -q, --quiet

    Decrease verbosity (can be repeated).

  -h, --help

    Show this message and exit.
"""

# Standard library modules.
import getopt
import json
import logging
import sys

# External dependencies.
import coloredlogs
from humanfriendly import (
    format_size,
    format_timespan,
    parse_size,
    parse_timespan,
    pluralize,
)
from humanfriendly.terminal import HIGHLIGHT_COLOR, ansi_wrap, output, usage, warning

# Modules included in our package.
from apache_manager import ApacheManager, NATIVE_WORKERS_LABEL
from apache_manager.interactive import watch_metrics

# Initialize a logger for this program.
logger = logging.getLogger(__name__)


def main():
    """Command line interface for the ``apache-manager`` program."""
    # Configure logging output.
    coloredlogs.install(syslog='notice')
    # Command line option defaults.
    actions = set()
    kw = dict()
    data_file = '/tmp/apache-manager.txt'
    dry_run = False
    # Parse the command line options.
    try:
        options, arguments = getopt.getopt(sys.argv[1:], 'ckwa:i:t:T:f:znvqh', [
            'collect-metrics', 'kill-workers', 'watch', 'max-memory-active=',
            'max-memory-idle=', 'max-ss=', 'max-time=',
            'hanging-worker-threshold=', 'data-file=', 'zabbix-discovery',
            'dry-run', 'simulate', 'verbose', 'quiet', 'help',
        ])
        for option, value in options:
            if option in ('-c', '--collect-metrics'):
                actions.add('collect')
            elif option in ('-k', '--kill-workers'):
                actions.add('kill')
            elif option in ('-w', '--watch'):
                actions.add('watch')
            elif option in ('-a', '--max-memory-active'):
                kw['max_memory_active'] = parse_size(value, binary=True)
            elif option in ('-i', '--max-memory-idle'):
                kw['max_memory_idle'] = parse_size(value, binary=True)
            elif option in ('-t', '--max-ss', '--max-time'):
                kw['worker_timeout'] = parse_timespan(value)
            elif option in ('-T', '--hanging-worker-threshold'):
                kw['hanging_worker_threshold'] = parse_timespan(value)
            elif option in ('-f', '--data-file'):
                data_file = value
            elif option in ('-z', '--zabbix-discovery'):
                actions.add('discovery')
            elif option in ('-n', '--dry-run', '--simulate'):
                logger.info("Performing a dry run ..")
                dry_run = True
            elif option in ('-v', '--verbose'):
                coloredlogs.increase_verbosity()
            elif option in ('-q', '--quiet'):
                coloredlogs.decrease_verbosity()
            elif option in ('-h', '--help'):
                usage(__doc__)
                return
        if arguments:
            raise Exception("This program doesn't support any positional arguments")
    except Exception as e:
        warning("Error: %s!", e)
        sys.exit(1)
    manager = ApacheManager(**kw)
    try:
        # Execute the requested action(s).
        if 'kill' in actions:
            manager.kill_workers(dry_run=dry_run)
        if 'watch' in actions:
            watch_metrics(manager)
        if 'discovery' in actions:
            report_zabbix_discovery(manager)
        # Render a summary of monitoring metrics when no action was requested.
        if not actions and data_file != '-':
            for line in report_metrics(manager):
                if line_is_heading(line):
                    line = ansi_wrap(line, color=HIGHLIGHT_COLOR)
                output(line)
    except Exception:
        logger.exception("Encountered unexpected exception, aborting!")
        sys.exit(1)
    finally:
        if 'collect' in actions and (data_file == '-' or not dry_run):
            manager.save_metrics(data_file)


def report_metrics(manager):
    """Create a textual summary of Apache web server metrics."""
    lines = ["Server metrics:"]
    for name, value in sorted(manager.server_metrics.items()):
        if name in ('total_traffic', 'bytes_per_second', 'bytes_per_request'):
            value = format_size(value)
        elif name == 'cpu_load':
            value = '%.1f%%' % value
        elif name == 'uptime':
            value = format_timespan(value)
        name = ' '.join(name.split('_'))
        name = name[0].upper() + name[1:]
        lines.append(" - %s: %s" % (name, value))
    main_label = "main Apache workers" if manager.wsgi_process_groups else "Apache workers"
    report_memory_usage(lines, main_label, manager.memory_usage)
    for name, memory_usage in sorted(manager.wsgi_process_groups.items()):
        report_memory_usage(lines, "WSGI process group '%s'" % name, memory_usage)
    return lines


def report_memory_usage(lines, label, memory_usage):
    """Create a textual summary of Apache worker memory usage."""
    lines.append("")
    workers = pluralize(len(memory_usage), "worker")
    lines.append("Memory usage of %s (%s):" % (label, workers))
    lines.append(" - Minimum: %s" % format_size(memory_usage.min))
    lines.append(" - Average: %s" % format_size(memory_usage.average))
    lines.append(" - Maximum: %s" % format_size(memory_usage.max))


def report_zabbix_discovery(manager):
    """Enable Zabbix low-level discovery of WSGI application groups."""
    worker_groups = [NATIVE_WORKERS_LABEL] + sorted(manager.wsgi_process_groups.keys())
    output(json.dumps({'data': [{'{#NAME}': name} for name in worker_groups]}))


def line_is_heading(line):
    """Check whether a line of output generated by :func:`report_metrics()` should be highlighted as a heading."""
    return line.endswith(':')
